src/query/src/window_sort.rs

```diff
@@ -0,0 +1,3101 @@
+        // min_val = 1, array = [3, 2, 1, 0], end = 2
+        // min_val = 1, array = [3, 2, 0], end = 2
+        let end = bisect::<false>(&[array.clone()], &[min_val.clone()], &[*opt])?;
+        (start, end)
+    } else {
+        // min_val = 1, array = [1, 2, 3], start = 0
+        // min_val = 1, array = [0, 2, 3], start = 1
+        let start = bisect::<true>(&[array.clone()], &[min_val.clone()], &[*opt])?;
+        // max_val = 3, array = [1, 3, 4], end = 1
+        // max_val = 3, array = [1, 2, 4], end = 2
+        let end = bisect::<true>(&[array.clone()], &[max_val.clone()], &[*opt])?;
+        (start, end)
+    };
+
+    Ok((start, end - start))
+}
+
+macro_rules! array_iter_helper {
+    ($t:ty, $arr:expr) => {{
+        let typed = $arr.as_any().downcast_ref::<PrimitiveArray<$t>>().unwrap();
+        let iter = typed.iter().enumerate();
+        Box::new(iter) as Box<dyn Iterator<Item = (usize, Option<i64>)>>
+    }};
+}
+
+/// Compare with options, note None is considered as NULL here
+fn cmp_with_opts<T: Ord>(
+    a: &Option<T>,
+    b: &Option<T>,
+    opt: &Option<SortOptions>,
+) -> std::cmp::Ordering {
+    if let Some(opt) = opt {
+        if let (Some(a), Some(b)) = (a, b) {
+            if opt.descending {
+                b.cmp(a)
+            } else {
+                a.cmp(b)
+            }
+        } else {
+            match (opt.nulls_first, a.is_none()) {
+                (true, true) => std::cmp::Ordering::Less,
+                (true, false) => std::cmp::Ordering::Greater,
+                (false, true) => std::cmp::Ordering::Greater,
+                (false, false) => std::cmp::Ordering::Less,
+            }
+        }
+    } else {
+        a.cmp(b)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct SortedRunSet<N: Ord> {
+    /// sorted runs with batch corresponding to them
+    runs_with_batch: Vec<(DfRecordBatch, SucRun<N>)>,
+    /// sorted column from eval sorting expr
+    sort_column: SortColumn,
+}
+
+/// A struct to represent a successive run in the input iterator
+#[derive(Debug, Clone, PartialEq)]
+struct SucRun<N: Ord> {
+    /// offset of the first element in the run
+    offset: usize,
+    /// length of the run
+    len: usize,
+    /// first non-null value in the run
+    first_val: Option<N>,
+    /// last non-null value in the run
+    last_val: Option<N>,
+}
+
+impl SucRun<Timestamp> {
+    /// Get the time range of the run, which is [min_val, max_val + 1)
+    fn get_time_range(&self) -> Option<TimeRange> {
+        let start = self.first_val.min(self.last_val);
+        let end = self
+            .first_val
+            .max(self.last_val)
+            .map(|i| Timestamp::new(i.value() + 1, i.unit()));
+        start.zip(end).map(|(s, e)| TimeRange::new(s, e))
+    }
+}
+
+/// find all successive runs in the input iterator
+fn find_successive_runs<T: Iterator<Item = (usize, Option<N>)>, N: Ord + Copy>(
+    iter: T,
+    sort_opts: &Option<SortOptions>,
+) -> Vec<SucRun<N>> {
```

Comment:
> We can assume the time index column is always non-null. This should simplify this function. Then we can use the `values()` to get a slice of i64. https://docs.rs/arrow/latest/arrow/array/struct.PrimitiveArray.html#method.values

This function is not complex anyway, and I want to make it more useful in that it can also accept null column in case future changes are needed here and we forget this invariant assumption. Also it's a lot of unnecessary chore to modify all unit tests for it